File: Leetcode/1608. Special Array With X Elements Greater Than or Equal X/solution.go
import (
	"math"
)

func specialArray(nums []int) int {
	var n = len(nums)
	var freq = make([]int, n+1)
	var numbersBiggerEqual = 0

	for _, num := range nums {
		freq[int(math.Min(float64(n), float64(num)))] += 1
	}

	for i := n; i >= 0; i-- {
		numbersBiggerEqual += freq[i]
		if numbersBiggerEqual == i {
			return i
		}
	}

	return -1
}

File: Leetcode/1608. Special Array With X Elements Greater Than or Equal X/solution.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:
        N = len(nums)
        freq = [0] * (N + 1) # The extra spaces will count for all the numbers bigger than N

        for num in nums:
            freq[min(N, num)] += 1
        
        numbers_bigger_equal = 0

        for i in range(N, 0, -1):
            numbers_bigger_equal += freq[i]
            if i == numbers_bigger_equal:
                return i
        
        return -1